### ISSUE TITLE
High CPU load when moving window between monitors

### ISSUE THREAD

**Issue by NickeZ** [jwilm/alacritty#1790]
Which operating system does the issue occur on?
Linux (kernel version 4.15)

If on linux, are you using X11 or Wayland?
X11, I3 (I don't have any compositor)

I've noticed recently that Alacritty sometimes can cause high cpu load. The total CPU usage of X11, I3 and Alacritty go to approximately 100% (i.e. fully utilize one core). It goes away when I exit the Alacritty process that is causing high load. I don't know exactly when it happens, sometimes it affects a terminal that I only launch in the background and basically haven't even used.

Is there anything I can do to debug it when it happens?

edit: 
```
$ git describe
v0.2.2-4-ga68ce7a
```

**Comment by chrisduerr** [comment]
Please let me know if you find a way to reliably reproduce this issue. I haven't run into this myself.

Restarting i3 with `i3-msg restart` would at least make sure there's no problem with i3. Restarting X is more complicated, since that would kill Alacritty.

Running it in full logging mode would probably help. The #1188 PR which logs to a file should help with this.

**Comment by HerrmannM** [comment]
Maybe this can help.

Which operating system does the issue occur on?
Linux 4.19.1-arch1-1-ARCH

If on linux, are you using X11 or Wayland?
X, XFCE 4.12

Running alacritty 0.2.3 (archlinux packaged)

I have a laptop connected to an external screen (both screens are 1920x1080).
When I have alacritty on my laptop screen (huge font, looks like the DPI thing is not working here), nothing wrong with the CPU (Xorg around 0.3%). When I move the alacritty window to the external screen (normal font, looks like the DPI thing is working here), the CPU usage of Xorg jumps to around 96%.

Thanks for the work!

**Comment by NickeZ** [comment]
I'll enable full logging when that is merged. For now it seems to happen rarely and I don't want to start all my terminals from other terminals just to get the output.

**Comment by EruIluvatar** [comment]
Which operating system does the issue occur on?
Solus, Linux 4.18.16-97.current

If on linux, are you using X11 or Wayland?
X11, i3 (no compositor / DE)

Running alacritty 0.2.3 (built from tagged release)

Same problem, also related to external displays.
I have 2 external displays and the same problem occurs if I switch from external -> internal and vice versa. Also the scaling is messed up with huge letters on the internal screen if switching external -> internal. `i3-msg restart` doesn't change anything.

external:
![external](https://user-images.githubusercontent.com/5493667/48605978-959c4e00-e97e-11e8-8c63-91f249f77912.png)

internal:
![internal](https://user-images.githubusercontent.com/5493667/48605977-959c4e00-e97e-11e8-9cef-e91e0842c158.png)

**Comment by chrisduerr** [comment]
@NickeZ The PR has been merged and it should allow you to enable full logging for all Alacritty sessions now.

**Comment by DenLilleMand** [comment]
Just if someone sits on `archlinux` and has the same problem, installing the `AIUR` community package `alacritty-git` fixes the problem. So seems like it is the package in the official repository that has a problem.

**Comment by chrisduerr** [comment]
@NickeZ Can you confirm that the latest master resolves this problem? It would kinda surprise me since I didn't really change anything related to CPU usage. Unless it was related to a dependency.

**Comment by NickeZ** [comment]
I didn't upgrade yet, and I haven't reproduced it yet. But I'll upgrade now and report back in a few days

**Comment by jonasrogert** [comment]
I have the same issue using arch, i3, alacritty-git from AUR (updated a few seconds ago) 0.2.3. 

Moving alacritty to another display will make xorg, i3 and alacritty to use 100% cpu. I have scaled one of the display as it is high dpi. I didn't have this problem a few days ago.

Anything i could provide to make the debug easier?

**Comment by DenLilleMand** [comment]
> I have the same issue using arch, i3, alacritty-git from AUR (updated a few seconds ago) 0.2.3.
> 
> Moving alacritty to another display will make xorg, i3 and alacritty to use 100% cpu. I have scaled one of the display as it is high dpi. I didn't have this problem a few days ago.
> 
> Anything i could provide to make the debug easier?

Isn't the scale high dpi setting deprecated? try to remove it from your `alacritty.yml`

**Comment by chrisduerr** [comment]
@jonasrogert Thanks for pointing out that this happens when moving windows across displays. I was able to reproduce this issue myself.

The issue is indeed resolved once the window is closed. I'll have to investigate if this is caused by Alacritty or a dependency of it.

**Comment by jonasrogert** [comment]
> > I have the same issue using arch, i3, alacritty-git from AUR (updated a few seconds ago) 0.2.3.
> > Moving alacritty to another display will make xorg, i3 and alacritty to use 100% cpu. I have scaled one of the display as it is high dpi. I didn't have this problem a few days ago.
> > Anything i could provide to make the debug easier?
> 
> Isn't the scale high dpi setting deprecated? try to remove it from your `alacritty.yml`

@DenLilleMand The scaling is done via x so the whole display is a lower resolution (using xrandr -scale 0.5x0.5.

**Comment by chrisduerr** [comment]
I didn't actually see that NickeZ reported this issue upstream with winit: https://github.com/tomaka/winit/issues/706.

I've investigated the problem myself and everything seems to point to winit as the culprit here. So this issue is blocked on a fix to winit.

**Comment by NickeZ** [comment]
I can also confirm that my issue originates from moving alacritty between monitors.

**Comment by dzmien** [comment]
I am having the same problem:
Which operating system does the issue occur on?
4.19.2-arch1-1-ARCH

If on linux, are you using X11 or Wayland?
X11, i3, xcompmgr

Running alacritty 0.2.3 from official Arch package. Moving alacritty between screens causes xorg's CPU usage to spike and plateau there until alacritty is closed. Restarting i3 in place does not fix the problem. Manipulating the alacritty window in any other way as long as it stays on the same monitor does not cause any problems. Interestingly, I can float the alacritty window and drag it to the other monitor and then push it back into tiling without issue.

Here is alacritty -vvv output starting from the command to move Alacritty to the other monitor (ends when I exit alacritty)

[11:00:25] Term::resize
[11:00:25] num_cols, num_lines = 118, 63
[11:00:25] width: 954, height: 1020
[11:00:25] carriage_return
[11:00:25] carriage_return
[11:00:25] move_up: 1
[11:00:25] goto: line=0, col=0
[11:00:25] Set Attribute: Reset
[11:00:25] Set Attribute: CancelReverse
[11:00:25] Set Attribute: CancelUnderline
[11:00:25] clear_screen: Below
[11:00:25] Set Attribute: Bold
[11:00:25] Set Attribute: Foreground(Named(Green))
[11:00:25] Set Attribute: Foreground(Named(Blue))
[11:00:25] Set Attribute: Reset
[11:00:25] Set Attribute: Foreground(Named(White))
[11:00:25] Set Attribute: Reset
[11:00:25] carriage_return
[11:00:25] linefeed
[11:00:25] Set Attribute: Foreground(Named(Blue))
[11:00:25] Set Attribute: Bold
[11:00:25] Set Attribute: Foreground(Named(Blue))
[11:00:25] Set Attribute: Reset
[11:00:25] Font size changed: Size(24) [DPR: 1]
[11:00:25] Term::resize
[11:00:25] Term::resize dimensions unchanged
[11:00:25] width: 954, height: 1020
[11:00:39] backspace
[11:00:41] unset_mode: CursorKeys
[11:00:41] unset mode::TermMode::APP_KEYPAD
[11:00:41] unset_mode: BracketedPaste
[11:00:41] carriage_return
[11:00:41] carriage_return
[11:00:41] linefeed
[11:00:41] Goodbye.

**Comment by 15cm** [comment]
Inspired by @dzmien, I find a temporary workaround is to toggle floating twice of all Alacritty instances every time we move Alacritty between monitors, using script like this:
```bash
for i in $(seq 2); do
  i3-msg '[class="Alacritty"] floating toggle'
done
```

**Comment by Jasu** [comment]
[I opened a PR fixing this on winit.](https://github.com/tomaka/winit/pull/737)

The temporary fix by @15cm works, since the problem is caused by `winit` attempting to preserve the DPI-adjusted window size when moving across monitors. Making Alacritty float allows the resize to the desired size pass, stopping `winit` from continuosly attempting to resize the window.

**Comment by adambyrtek** [comment]
The underlying issue seems to be fixed, so hope it can be included in Alacritty soon, but just for the record I'm experiencing a similar same issue with Gnome Shell 3.28, Xorg 7.7, and Alacrity 0.2.3.

Sometimes when plugging an external monitor CPU usage of Xorg jumps to 40% and Gnome Shell to 50%. Alacritty hovers around 7% but everything gets back to normal if I kill Alacritty.

**Comment by chrisduerr** [comment]
@adambyrtek The issue is not fixed yet.

**Comment by chrisduerr** [comment]
As soon as winit pushes a new (non-breaking) update, we can update our deps and this should fix itself.

Thanks to the winit crowd for resolving this issue!

**Comment by jypma** [comment]
I have the same symptom when changing resolution through `xrandr` while a (highdpi enabled) alacritty is open. Would the `winit` fix also work in that situation?

**Comment by chrisduerr** [comment]
I'd say it's very likely to fix that issue too. If it's not fixed in the next release, you can open a separate issue though.

**Comment by FichteFoll** [comment]
I noticed this issue too after not updating alacritty for probably 2 months and then testing the underline fixes/PR. However, I **don't** have to move the window to a different monitor for this to occur. 

It usually happens with my `htop` terminal that I just launch shortly after booting. I open it already on my second monitor and don't ever move the window afterwards. It can, however, get resized when I add another window to that desktop, but the issue surfaces regardless.

I am running Arch and i3 (and X11).

![2019-01-03_06-28-50](https://user-images.githubusercontent.com/931051/50806151-c1d8a900-12f5-11e9-9e1c-fd6035ef1b02.png)
![2019-01-03_06-29-13](https://user-images.githubusercontent.com/931051/50806152-c1d8a900-12f5-11e9-9c64-4265ceee182c.png)

**Comment by frostRed** [comment]
![cpu1](https://user-images.githubusercontent.com/2486301/51448700-4ea64c80-1d63-11e9-8fe5-bcb684a3b2ef.png)
![cpu2](https://user-images.githubusercontent.com/2486301/51448703-549c2d80-1d63-11e9-8b94-9fc02d8ebd59.png)

This problem still exists in Manjaro.

```
LSB Version:    n/a
Distributor ID: ManjaroLinux
Description:    Manjaro Linux
Release:        18.0.2
Codename:       Illyria
 
alacritty --version
alacritty 0.2.5
```
